pkg/rtc/uptrackmanager.go

```diff
@@ -0,0 +1,664 @@
+}
+
+// should be called with lock held
+func (u *UptrackManager) getPublishedTrack(sid string) types.PublishedTrack {
+	return u.publishedTracks[sid]
+}
+
+// should be called with lock held
+func (u *UptrackManager) getPublishedTrackBySignalCid(clientId string) types.PublishedTrack {
+	for _, publishedTrack := range u.publishedTracks {
+		if publishedTrack.SignalCid() == clientId {
+			return publishedTrack
+		}
+	}
+
+	return nil
+}
+
+// should be called with lock held
+func (u *UptrackManager) getPublishedTrackBySdpCid(clientId string) types.PublishedTrack {
+	for _, publishedTrack := range u.publishedTracks {
+		if publishedTrack.SdpCid() == clientId {
+			return publishedTrack
+		}
+	}
+
+	return nil
+}
+
+// should be called with lock held
+func (u *UptrackManager) getPendingTrack(clientId string, kind livekit.TrackType) (string, *livekit.TrackInfo) {
+	signalCid := clientId
+	trackInfo := u.pendingTracks[clientId]
+
+	if trackInfo == nil {
+		//
+		// If no match on client id, find first one matching type
+		// as MediaStreamTrack can change client id when transceiver
+		// is added to peer connection.
+		//
+		for cid, ti := range u.pendingTracks {
+			if ti.Type == kind {
+				trackInfo = ti
+				signalCid = cid
+				break
+			}
+		}
+	}
+
+	// if still not found, we are done
+	if trackInfo == nil {
+		u.params.Logger.Errorw("track info not published prior to track", nil, "clientId", clientId)
+	}
+	return signalCid, trackInfo
+}
+
+func (u *UptrackManager) handleTrackPublished(track types.PublishedTrack) {
+	u.lock.Lock()
+	if _, ok := u.publishedTracks[track.ID()]; !ok {
+		u.publishedTracks[track.ID()] = track
+	}
+	u.lock.Unlock()
+
+	track.AddOnClose(func() {
+		// cleanup
+		u.lock.Lock()
+		trackSid := track.ID()
+		delete(u.publishedTracks, trackSid)
+		delete(u.pendingSubscriptions, trackSid)
+		delete(u.subscriptionPermissions, trackSid)
+		u.lock.Unlock()
+		// only send this when client is in a ready state
+		if u.onTrackUpdated != nil {
+			u.onTrackUpdated(track, true)
+		}
+	})
+
+	if u.onTrackPublished != nil {
+		u.onTrackPublished(track)
+	}
+}
+
+func (u *UptrackManager) updateSubscriptionPermissions(permissions *livekit.UpdateSubscriptionPermissions) {
+	// every update overrides the existing
+	u.subscriptionPermissions = make(map[string][]string)
+
+	// all_participants takes precedence
+	if permissions.AllParticipants {
+		// everything is allowed, nothing else to do
+		return
+	}
+
+	// if track_permissions is empty, nobody is allowed to subscribe to any tracks
+	if permissions.TrackPermissions != nil && len(permissions.TrackPermissions) == 0 {
+		for trackSid, _ := range u.publishedTracks {
+			u.subscriptionPermissions[trackSid] = []string{}
+		}
+
+		for _, ti := range u.pendingTracks {
+			u.subscriptionPermissions[ti.Sid] = []string{}
+		}
+		return
+	}
+
+	// per participant permissions
+	for _, trackPerms := range permissions.TrackPermissions {
+		// all_tracks get precedence
+		if trackPerms.AllTracks {
```

Comment:
I think this would cause a problem after the publisher publishes another track after setting permissions.

A couple of options:
* store all options as a nil (i.e. delete(u.subscriptionPermissions, t.Sid)
* store the `TrackPermission` object as is, i.e. participantSid => `TrackPermission`, `hasPermission` could check by walking the list.

Reply:
Ah, good catch! Storing participant view of things is better. The protocol is designed that way. But, as most things have a track centric view, I was doing the dance to convert that. Will use the participant centric view in code which will make this easier and also address the case you mentioned.

Reply:
@davidzhao Addressing this and also cleaning up pending subscriptions in this commit - https://github.com/livekit/livekit-server/pull/284/commits/dbab1e819db889640ec3d4b21af6cfc655a5a380. (have a typo in a comment :-) which I will fix in a subsequent commit as I want to add a test for new track after permissions set).

The down track double close we discussed is not an issue I think because the disallowed subscription will not in the subscribed list and hence not going to call down track close.

DownTrack.Close() has a couple of method outside the close once (setting forwarder mute and sending blank frames). Just for safety, maybe we add something to prevent double close?

Reply:
Also, chose sending back `disallowedSubscribers` in `OnClose` callback of `Participant` as quite a few places was calling participant close (including from within the participant when ICE failed). So, to reduce too many changes, the OnClose change made sense. Let me know if there is a better way.